complex/src/scalar.rs

```diff
@@ -0,0 +1,252 @@
+
+use traits::{Num, FromPrimitive, Float};
+use std::ops::Neg;
+
+use Complex;
+
+pub trait Scalar: Num + Copy + Neg<Output = Self> {
+    /// Associated Real type
+    type Real;
+    /// Associated Complex type
+    type Complex;
+
+    /// Take the square root of a number.
+    fn sqrt(&self) -> Self;
+    /// Returns `e^(self)`, (the exponential function).
+    fn exp(&self) -> Self;
+    /// Returns the natural logarithm of the number.
+    fn ln(&self) -> Self;
+    /// Returns the square of the absolute value of the number
+    fn abs_sqr(&self) -> Self::Real;
+    /// Returns the absolute value of the number
+    fn abs(&self) -> Self::Real;
+
+    /// Raise a number to an integer power.
+    fn powi(&self, exp: i32) -> Self;
+    /// Raise a number to a floating point power.
+    fn powf(&self, exp: Self::Real) -> Self;
+    /// Raise a number to a complex power.
+    fn powc(&self, exp: Self::Complex) -> Self::Complex;
+
+    /// Returns complex-conjugate number
+    fn conj(&self) -> Self;
+
+    /// Computes the sine of a number
+    fn sin(&self) -> Self;
+    /// Computes the cosine of a number
+    fn cos(&self) -> Self;
+    /// Computes the tangent of a number
+    fn tan(&self) -> Self;
+    /// Computes the arcsine of a number
+    fn asin(&self) -> Self;
+    /// Computes the arccosine of a number
+    fn acos(&self) -> Self;
+    /// Computes the arctangent of a number
+    fn atan(&self) -> Self;
+
+    /// Computes the hyperbolic-sine of a number
+    fn sinh(&self) -> Self;
+    /// Computes the hyperbolic-cosine of a number
+    fn cosh(&self) -> Self;
+    /// Computes the hyperbolic-tangent of a number
+    fn tanh(&self) -> Self;
+    /// Computes the hyperbolic-arcsine of a number
+    fn asinh(&self) -> Self;
+    /// Computes the hyperbolic-arccosine of a number
+    fn acosh(&self) -> Self;
+    /// Computes the hyperbolic-arctangent of a number
+    fn atanh(&self) -> Self;
+
+    /// Checks if the given (real or imaginary part of complex) number is NaN
+    fn is_nan(self) -> bool;
+    /// Checks if the given (real or imaginary part of complex) number is infinite
+    fn is_infinite(self) -> bool;
+    /// Checks if the given number is finite
+    fn is_finite(self) -> bool;
+    /// Checks if the given number is normal
+    fn is_normal(self) -> bool;
+}
+
+impl<T: Clone + Float + FromPrimitive> Scalar for Complex<T> {
```

Comment:
Hmm, it's a little awkward that `FromPrimitive` is needed just for `powi`.

Reply:
In my experience using `ndarray_linalg::Scalar`, the original of this PR, `Scalar::from_f64` is necessary.
I would like to add `trait Scalar: FromPrimitive + ...` restriction, but `Complex` does not implement it.